apps/nsq_to_nsq/nsq_to_nsq.go

```diff
@@ -301,6 +285,81 @@ func main() {
 	if len(destNsqdTCPAddrs) == 0 {
 		log.Fatal("--destination-nsqd-tcp-address required")
 	}
+}
+
+
+func initConsumerAndHandler(cCfg *nsq.Config,
+	                        producers map[string]*nsq.Producer,
+	                        selectedMode int,
+	                        hostPool hostpool.HostPool,
+	                        perAddressStatus map[string]*timer_metrics.TimerMetrics) []*nsq.Consumer {
+	var consumerList []*nsq.Consumer
+	var singleDestinationHandler *PublishHandler
+	var handlerList []*PublishHandler
+
+	if *destTopic != "" {
+		singleDestinationHandler = &PublishHandler{
+			addresses:        destNsqdTCPAddrs,
+			producers:        producers,
+			mode:             selectedMode,
+			hostPool:         hostPool,
+			respChan:         make(chan *nsq.ProducerTransaction, len(destNsqdTCPAddrs)),
+			perAddressStatus: perAddressStatus,
+			timermetrics:     timer_metrics.NewTimerMetrics(*statusEvery, "[aggregate]:"),
+			destinationTopic: *destTopic,
+		}
+		handlerList = append(handlerList, singleDestinationHandler)
+	}
+
+	for _, topic := range topics {
+		consumer, err := nsq.NewConsumer(topic, *channel, cCfg)
+		consumerList = append(consumerList, consumer)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if (singleDestinationHandler != nil) {
+			consumer.AddConcurrentHandlers(singleDestinationHandler, len(destNsqdTCPAddrs))
+		} else {
+			handler := &PublishHandler{
+				addresses:        destNsqdTCPAddrs,
+				producers:        producers,
+				mode:             selectedMode,
+				hostPool:         hostPool,
+				respChan:         make(chan *nsq.ProducerTransaction, len(destNsqdTCPAddrs)),
+				perAddressStatus: perAddressStatus,
+				timermetrics:     timer_metrics.NewTimerMetrics(*statusEvery, "[aggregate]:"),
+				destinationTopic: topic,
+			}
+			consumer.AddConcurrentHandlers(handler, len(destNsqdTCPAddrs))
+			handlerList = append(handlerList, handler)
+		}
+	}
+	for i := 0; i < len(destNsqdTCPAddrs); i++ {
+		for _, handler := range handlerList {
+			go handler.responder()
+		}
+	}
+
+	return consumerList;
+}
+
+func main() {
+	var selectedMode int
+
+	cCfg := nsq.NewConfig()
+	pCfg := nsq.NewConfig()
+
+	flag.Var(&nsq.ConfigFlag{cCfg}, "consumer-opt", "option to passthrough to nsq.Consumer (may be given multiple times, see http://godoc.org/github.com/nsqio/go-nsq#Config)")
+	flag.Var(&nsq.ConfigFlag{pCfg}, "producer-opt", "option to passthrough to nsq.Producer (may be given multiple times, see http://godoc.org/github.com/nsqio/go-nsq#Config)")
+
+	flag.Parse()
+
+	if *showVersion {
+		fmt.Printf("nsq_to_nsq v%s\n", version.Binary)
+		return
+	}
+
+	commandLineValidation()
```

Comment:
moving some code into this new function above main() just increases the diff, without reducing code duplication or improving structure